internal/executor/utilisation/cluster_utilisation.go

```diff
@@ -165,24 +172,63 @@ func (clusterUtilisationService *ClusterUtilisationService) GetAllNodeGroupAlloc
 		return []*NodeGroupAllocationInfo{}, err
 	}
 
+	batchPods, err := clusterUtilisationService.clusterContext.GetBatchPods()
+	if err != nil {
+		// Failing this step will just mean cordoned utilisation won't be accounted for
+		// This isn't a critical failure so we log and continue
+		log.Errorf("Error fetching batch pods to calculate cordond node utilisation: %v", err)
+	}
+
 	nodeGroups := clusterUtilisationService.nodeInfoService.GroupNodesByType(allAvailableProcessingNodes)
 	result := make([]*NodeGroupAllocationInfo, 0, len(nodeGroups))
 
 	for _, nodeGroup := range nodeGroups {
 		totalNodeResource := common.CalculateTotalResource(nodeGroup.Nodes)
 		allocatableNodeResource := allocatableResourceByNodeType[nodeGroup.NodeType.Id]
+		cordonedNodeResource := clusterUtilisationService.getCordonedResource(nodeGroup.Nodes, batchPods)
 
 		result = append(result, &NodeGroupAllocationInfo{
 			NodeType:                     nodeGroup.NodeType,
 			Nodes:                        nodeGroup.Nodes,
 			NodeGroupCapacity:            totalNodeResource,
 			NodeGroupAllocatableCapacity: allocatableNodeResource,
+			NodeGroupCordonedCapacity:    cordonedNodeResource,
 		})
 	}
 
 	return result, nil
 }
 
+// getCordonedResource takes a list of nodes and a list of pods and returns the resources allocated
+// to pods running on cordoned nodes. We need this information in calculating queue fair shares when
+// significant resource is running on cordoned nodes.
+func (clusterUtilisationService *ClusterUtilisationService) getCordonedResource(nodes []*v1.Node, pods []*v1.Pod) common.ComputeResources {
+	cordonedNodes := util.FilterNodes(nodes, func(node *v1.Node) bool { return node.Spec.Unschedulable })
+	podsOnNodes := GetPodsOnNodes(pods, cordonedNodes)
+	usage := common.ComputeResources{}
+	for _, pod := range podsOnNodes {
+		for _, container := range pod.Spec.Containers {
+			containerResource := castAsComputeResources(container.Resources.Limits) // Not 100% on whether this should be Requests or Limits
+			usage.Add(containerResource)
```

Comment:
so I don't know nearly enough about k8s here, but there' no risk of overwriting keys here is there?

Reply:
I'm fairly certain there is not, especially with the change "castAsComputeResources->FromResourceList" which explicitly copies containers.

I'm not entirely sure what you're thinking would override keys or what keys.

Reply:
so  `usage` here is essentially a map of `string` to `Resource.Quantity`.  I was wondering if calling `Add` on the map would do the correct thing in terms of adding keys together rather than replacing previously added keys with the latest one. I checked the code and it does do the correct thing so I think we're good.